core/features/feature_engineering.py

```diff
@@ -0,0 +1,400 @@
+    mapping_dict = {old_col: new_col for old_col, new_col in zip(old_cols, new_cols)}
+    df_dummies_comb = df_dummies_comb.rename(mapping_dict)
+
+    # Sort the columns in the logical order
+    df_dummies_comb = df_dummies_comb[col_order]
+
+    # Join the dummy columns with the original dataframe
+    df_res = pl.concat([df, df_dummies_comb], how="horizontal")
+
+    logger.info("Finished creating LUI columns.")
+
+    return df_res
+
+
+def group_land_use_types_and_intensities(
+    df: pl.DataFrame, col_order: list[str]
+) -> pl.DataFrame:
+    """
+    Creates certain special groupings of land-use types and intensities:
+    1. Urban land use of all intensities are combined into one column.
+    2. All secondary vegetation types are combined into one column, but with
+    different levels of intensity.
+    3. Plantation forest is grouped with secondary vegetation.
+
+    Args:
+        df: Dataframe containing 'Predominant_land_use' and 'Use_intensity'
+            columns.
+        col_order: The order of the combined dummies in the output df.
+
+        Returns:
+            df: Updated df with the special groupings added.
+    """
+    # Combine urban land use of all intensities
+    df = df.with_columns(
+        pl.when(pl.col("Predominant_land_use") == "Urban")
+        .then(1)
+        .otherwise(0)
+        .alias("Urban_All uses")
+    )
+
+    # Combine all secondary vegetation types
+    def _secondary_veg_intensity(row: pl.String) -> str:
+        if "secondary" in str(row).lower():
+            new_row = "Secondary vegetation_" + str(row).split("_")[1]
+            return new_row
+        else:
+            return row
+
+    df = df.with_columns(
+        pl.col("LU_type_intensity")
+        .map_elements(lambda row: _secondary_veg_intensity(row))
+        .alias("Secondary_veg_intensity")
+    )
+
+    # Group plantation forest with secondary vegetation
+    df = df.with_columns(
+        pl.when(pl.col("Secondary_veg_intensity").str.contains("(?i)plantation"))
+        .then(
+            pl.when(pl.col("Secondary_veg_intensity").str.contains("(?i)minimal"))
+            .then(pl.lit("Secondary vegetation_Light use"))
+            .otherwise(pl.lit("Secondary vegetation_Intense use"))
+        )
+        .otherwise(pl.col("Secondary_veg_intensity"))
+        .alias("Secondary_veg_intensity")
+    )
+
+    # Create dummy columns from the combined secondary vegetation column
+    df_dummies = df.select("Secondary_veg_intensity").to_dummies(
+        "Secondary_veg_intensity"
+    )
+
+    old_cols = df_dummies.columns
+    new_cols = [col.replace("Secondary_veg_intensity_", "") for col in old_cols]
+    mapping_dict = {old_col: new_col for old_col, new_col in zip(old_cols, new_cols)}
+    df_dummies = df_dummies.rename(mapping_dict)
+
+    df_dummies = df_dummies[col_order]
+    df = pl.concat([df, df_dummies], how="horizontal")
+
+    # Combine light and intense use for cropland and pasture
+    df = df.with_columns(
```

Comment:
With so many dataframe operations going on in some of these operations I think it's important to having good variable naming, to avoid any accidental dataframe operations or overwrites. I think generally you've got a good handle on this, but for instance the generic `df` can get a bit confusing by reading through the end of some of these functions. There's obviously tradeoff between updating the naming based on current operations, because then you're holding more dataframe references in memory and all that, but, just a thought!

Reply:
Also a good point, and something I'm guilty of in many place. Added to the backlog too!